File: 7.3/replaceBlank.cpp

#include<iostream>
using namespace std;

void replaceBlank(char str[],int length){
   if(str==NULL||length<=0) return;
   int originLength=0;
   int blankNum=0;
   int i=0;
   for(i=0;str[i]!='\0';i++){
    originLength++;
    if(str[i]==' '){
        blankNum++;
    }
   }
   int newLength=originLength+blankNum*2;
   if(newLength>length) return;

   while(originLength>=0&&newLength>originLength){
    //当字符串没有空格时，就会出现ewLength=originLength的情况，从而减少复制操作
    if(str[originLength]==' '){
        str[newLength--]='0';
        str[newLength--]='2';
        str[newLength--]='%';
    }else{
        str[newLength--]=str[originLength];
    }
    originLength--;
   }
   cout<<str<<endl;
}

int main(){
const int length=100;
char a[length]="hello world!";
replaceBlank(a,length);
}

File: 7.3/replaceBlank.js
const replaceBlank = (str, length) => {
    if (str === null && length <= 0) {
        return 0;
    }
    let orgLength = str.length;
    let blankNum = 0;

    for (let i = 0; i < str.length; i++) {
        if (str[i] === ' ') {
            blankNum++;
        }
    }

    let newLength = orgLength + blankNum * 2;

    if (newLength > length) return;

    while(orgLength>=0&&newLength>orgLength){
        //当字符串没有空格时，就会出现ewLength=originLength的情况，从而减少复制操作
        if(str[orgLength]===' '){
            str[newLength--]='0';
            str[newLength--]='2';
            str[newLength--]='%';
        }else{
            str[newLength--]=str[orgLength];
        }
        orgLength--;
    }
    return str;
};

const main = () => {
    let length = 100;
    let str = 'wei qiu juan';
    console.log(replaceBlank(str, length));
};

main();//哈哈哈，没跑出来，我也不知道为啥